flow-server/src/test/java/com/vaadin/flow/server/frontend/NodeUpdatePackagesNpmVersionLockingTest.java

```diff
@@ -0,0 +1,174 @@
+    private static final String USER_PINNED_DEPENDENCY_VERSION = "1.0";
+
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    private File baseDir;
+    private File generatedDir;
+    private File resourcesDir;
+
+    private ClassFinder classFinder;
+
+    private FeatureFlags featureFlags;
+
+    @Before
+    public void setup() throws Exception {
+        baseDir = temporaryFolder.getRoot();
+
+        featureFlags = Mockito.mock(FeatureFlags.class);
+
+        generatedDir = new File(baseDir,
+                Paths.get(TARGET, DEFAULT_GENERATED_DIR).toString());
+        resourcesDir = new File(baseDir,
+                Paths.get(TARGET, DEFAULT_FLOW_RESOURCES_FOLDER).toString());
+
+        FrontendStubs.createStubNode(true, true, baseDir.getAbsolutePath());
+
+        classFinder = Mockito.spy(getClassFinder());
+        File versions = temporaryFolder.newFile();
+        FileUtils.write(versions,
+                String.format(
+                        "{" + "\"vaadin-overlay\": {"
+                                + "\"npmName\": \"@vaadin/vaadin-overlay\","
+                                + "\"jsVersion\": \"%s\"" + "}" + "}",
+                        PLATFORM_PINNED_DEPENDENCY_VERSION),
+                StandardCharsets.UTF_8);
+        // @formatter:on
+
+        Mockito.when(classFinder.getResource(Constants.VAADIN_VERSIONS_JSON))
+                .thenReturn(versions.toURI().toURL());
+    }
+
+    @Test
+    public void shoudlLockPinnedVersion_whenExistsInDependencies()
+            throws IOException {
+        TaskUpdatePackages packageUpdater = createPackageUpdater();
+        JsonObject packageJson = packageUpdater.getPackageJson();
+        packageJson.getObject(DEPENDENCIES).put(TEST_DEPENDENCY,
+                PLATFORM_PINNED_DEPENDENCY_VERSION);
+        Assert.assertNull(packageJson.getObject(OVERRIDES));
+
+        String versionsPath = packageUpdater.generateVersionsJson();
+        packageUpdater.lockVersionForNpm(packageJson, versionsPath);
+
+        Assert.assertEquals("$" + TEST_DEPENDENCY,
+                packageJson.getObject(OVERRIDES).getString(TEST_DEPENDENCY));
+    }
+
+    @Test
+    public void shoudlNotLockPinnedVersion_whenNotExistsInDependencies()
+            throws IOException {
+        TaskUpdatePackages packageUpdater = createPackageUpdater();
+        JsonObject packageJson = packageUpdater.getPackageJson();
+
+        Assert.assertNull(packageJson.getObject(OVERRIDES));
+        Assert.assertNull(
+                packageJson.getObject(DEPENDENCIES).get(TEST_DEPENDENCY));
+
+        String versionsPath = packageUpdater.generateVersionsJson();
+        packageUpdater.lockVersionForNpm(packageJson, versionsPath);
+
+        Assert.assertNull(
+                packageJson.getObject(OVERRIDES).get(TEST_DEPENDENCY));
+    }
+
+    @Test
+    public void shouldNotNotUpdatesOverrides_whenHasUserModification()
```

Comment:
```suggestion
    public void shouldNotUpdateOverrides_whenHasUserModification()
```
What is the expected place to have user modificatio? Override or dependencies? should override be populated for user dependency that is overriding framework version?

Reply:
the overrides section might have user modifications. A user could update the values in the overrides section manually, if there are such changes, we just leave them alone.

Reply:
If the update has pinned a Vaadin managed dep (e.g. updated in the `overides` section `"@polymer/iron-a11y-announcer": "$@polymer/iron-a11y-announcer"` -> `"@polymer/iron-a11y-announcer": "3.1.0"`), this is also cleaned. But maybe this is fine.

Reply:
right, it's an edge case and should be fine.